### PR TITLE
hardening/808_update_post_subscriptions_method

### PR DESCRIPTION
**Pull request by pcoello25**
* Update issue #808 
  * Update CHANGES_NEXT_RELEASE is not neccesary
* 100% Unit test passed:
```
Results :
[CYGNUS-COMMON] Tests run: 43, Failures: 0, Errors: 0, Skipped: 0
[CYGNUS-NGSI]   Tests run: 84, Failures: 0, Errors: 0, Skipped: 0
```
* e2e (unofficial) test passed:
```
curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {"entities": [{"idPattern": ".*","type": "Room"}],"condition": {"attrs": ["temperature"],"expression": {"q": "temperature>40"}}},"notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"expires": "2016-05-05T14:00:00.00Z","throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"true","result" : { SubscriptionID = 572ae23d20e1387832ed98d0}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"expires": "2016-05-05T14:00:00.00Z","throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'subject' is missing"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {"entities": [{"idPattern": ".*","type": "Room"}],"condition": {"attrs": ["temperature"],"expression": {"q": "temperature>40"}}},"expires": "2016-05-05T14:00:00.00Z","throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'notification' is missing"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {"entities": [{"idPattern": ".*","type": "Room"}]},"notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"expires": "2016-05-05T14:00:00.00Z","throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'condition' is missing"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {"condition": {"attrs": ["temperature"],"expression": {"q": "temperature>40"}}},"notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"expires": "2016-05-05T14:00:00.00Z","throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'entities' is missing"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {"entities": [{"idPattern": ".*","type": "Room"}],"condition": {"attrs": ["temperature"]}},"notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"expires": "2016-05-05T14:00:00.00Z","throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'subject' has missing fields"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {},"notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"expires": "2016-05-05T14:00:00.00Z","throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'subject' is missing"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {"entities": [{"idPattern": ".*","type": "Room"}],"condition": {"attrs": ["temperature"],"expression": {"q": "temperature>40"}}}, "expires": "2016-05-05T14:00:00.00Z","throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'notification' is missing"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {"entities": [{"idPattern": ".*","type": "Room"}],"condition": {"attrs": ["temperature"],"expression": {"q": "temperature>40"}}},"notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"expires": "2016-05-05T14:00:00.00Z","throttling": 5}}'
{"success":"false","error":"Missing endpoint"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"subject": {"entities": [{"idPattern": ".*","type": "Room"}],"condition": {"attrs": ["temperature"],"expression": {"q": "temperature>40"}}},"notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"expires": "2016-05-05T14:00:00.00Z","throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'description' is missing"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {"entities": [{"idPattern": ".*","type": "Room"}],"condition": {"attrs": ["temperature"],"expression": {"q": "temperature>40"}}},"notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"throttling": 5}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'expires' is missing"}

curl -X POST "http://localhost:8081/v1/subscriptions?ngsi_version=2" -d '{"subscription":{"description": "One subscription to rule them all","subject": {"entities": [{"idPattern": ".*","type": "Room"}],"condition": {"attrs": ["temperature"],"expression": {"q": "temperature>40"}}},"notification": {"http": {"url": "http://localhost:1234"},"attrs": ["temperature","humidity"]},"expires": "2016-05-05T14:00:00.00Z"}, "endpoint":{"host":"orion.lab.fiware.org", "port":"1026", "ssl":"false", "xauthtoken":"QsENv67AJj7blC2qJ0YvfSc5hMWYrs"}}'
{"success":"false","error":"Invalid subscription, field 'throttling' is missing"}
```
* Assignee: @frbattid . In addition, comments of @fgalan are welcome :)

### CODE REVIEW

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/orion/OrionBackend.java [hunk 1]
```diff
@@ -83,4 +91,4 @@ public JsonResponse deleteSubscriptionV2(String subscriptionId, String token)
     public JsonResponse getSubscriptionsByIdV2(String token, 
             String subscriptionId) throws Exception;
     
-} // StatsBackend
+} // StatsBackend
```
Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/orion/OrionBackend.java]: Extra final line required by Github :)

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/orion/OrionBackendImpl.java [hunk 2]
```diff
@@ -186,4 +200,4 @@ public void updateContext(String entityId, String entityType, ArrayList<OrionSta
         } // if
     } // updateContext
     **/
-} // StatsBackendImpl
+} // StatsBackendImpl
```
Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/orion/OrionBackendImpl.java]: Extra final line required by Github :)

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/management/ManagementInterface.java [hunk 3]
```diff
@@ -15,6 +15,23 @@
  *
  * For those usages not covered by the GNU Affero General Public License please contact with iot_support at tid dot es
  */
+/**
```
Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/management/ManagementInterface.java]: The header is now duplicated :)

**Comment by frbattid** [comment]
Is the documentation (.md files and Apiary) updated? If not, it should be part of this PR.

**Comment by pcoello25** [comment]
Not yet, but it will be implemented (I forgot :tired_face: )

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/orion/OrionBackend.java (hunk 1 above)
Comment (pcoello25) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/orion/OrionBackend.java]: Fixed in c4c29ba

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/orion/OrionBackendImpl.java (hunk 2 above)
Comment (pcoello25) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/orion/OrionBackendImpl.java]: Fixed in c4c29ba

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/management/ManagementInterface.java (hunk 3 above)
Comment (pcoello25) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/management/ManagementInterface.java]: Fixed in c4c29ba

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/containers/CygnusSubscriptionV2.java [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+            default:
+                // Unreachable statement
+        } // switch
+      
+        LOGGER.debug("Valid input JSON.");
+        return 0;
+
+    } // isValid  
+    
+} // CygnusSubscriptionV1
```
Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/containers/CygnusSubscriptionV2.java]: A final white line is still necessary here

Comment (pcoello25) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/containers/CygnusSubscriptionV2.java]: Fixed in c591b71

**Comment by frbattid** [comment]
LGTM

**Pull request merged**